<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [3]:
import pandas as pd
import psycopg2

In [5]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [6]:
query_3_1 = f'''select count(*)
                from vacancies
            '''

In [7]:
qty_vac= pd.read_sql_query(query_3_1, connection)
display(qty_vac)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\1157217138.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  qty_vac= pd.read_sql_query(query_3_1, connection)


,count
0,49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [8]:
query_3_2 = f'''select count(*)
                from employers
            '''

In [9]:
qty_empr= pd.read_sql_query(query_3_2, connection)
display(qty_empr)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\2458544806.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  qty_empr= pd.read_sql_query(query_3_2, connection)


,count
0,23501


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [10]:
query_3_3 = f'''select count(*)
                from areas
            '''

In [11]:
qty_areas= pd.read_sql_query(query_3_3, connection)
display(qty_areas)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\1717760068.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  qty_areas= pd.read_sql_query(query_3_3, connection)


,count
0,1362


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [12]:
query_3_4 = f'''select count(*)
                from industries
            '''

In [13]:
qty_ind= pd.read_sql_query(query_3_4, connection)
display(qty_ind)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\408090445.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  qty_ind= pd.read_sql_query(query_3_4, connection)


,count
0,294


***

Начали работу с данными. Узнали общее количество вакансий, работодателей, регионов и сфер деятельности. Интересно, плчему так много регионов, возможно есть вакансии вне России. Можно также заметить, что в среднем на одного работодателя приходятся 2 вакансии.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [14]:
query_4_1 = f'''select a.name region, count(v.id) qty
                from vacancies v
                    join areas a on a.id=v.area_id
                group by a.name
                order by 2 desc
                limit 5
            '''

In [15]:
cnt_area= pd.read_sql_query(query_4_1, connection)
display(cnt_area)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\3807556044.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_area= pd.read_sql_query(query_4_1, connection)


,region,qty
0,Москва,5333
1,Санкт-Петербург,2851
2,Минск,2112
3,Новосибирск,2006
4,Алматы,1892


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [16]:
query_4_2 = f'''select count(*)
                from vacancies
                where salary_from is not null or salary_to is not null
            '''

In [17]:
qty_with_sal= pd.read_sql_query(query_4_2, connection)
display(qty_with_sal)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\2001132973.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  qty_with_sal= pd.read_sql_query(query_4_2, connection)


,count
0,24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [18]:
query_4_3 = f'''select round(avg(salary_from)) sf, round(avg(salary_to)) st
                from vacancies
            '''

In [19]:
avg_from_to= pd.read_sql_query(query_4_3,connection)
display(avg_from_to)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\729226904.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_from_to= pd.read_sql_query(query_4_3,connection)


,sf,st
0,71065.0,110537.0


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [20]:
query_4_4 = f'''select schedule, employment, count(*)
                from vacancies
                group by schedule, employment
                order by 3 desc
            '''

In [21]:
qty_pairs= pd.read_sql_query(query_4_4, connection)
display(qty_pairs)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\875815956.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  qty_pairs= pd.read_sql_query(query_4_4, connection)


,schedule,employment,count
0,Полный день,Полная занятость,35367
1,Удаленная работа,Полная занятость,7802
2,Гибкий график,Полная занятость,1593
3,Удаленная работа,Частичная занятость,1312
4,Сменный график,Полная занятость,940
5,Полный день,Стажировка,569
6,Вахтовый метод,Полная занятость,367
7,Полный день,Частичная занятость,347
8,Гибкий график,Частичная занятость,312
9,Полный день,Проектная работа,141


5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [22]:
query_4_4 = f'''select distinct experience, count(experience)
                from vacancies
                group by experience
                order by 2 asc
            '''

In [23]:
exp_cnt= pd.read_sql_query(query_4_4, connection)
display(exp_cnt)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\3426249347.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  exp_cnt= pd.read_sql_query(query_4_4, connection)


,experience,count
0,Более 6 лет,1337
1,Нет опыта,7197
2,От 3 до 6 лет,14511
3,От 1 года до 3 лет,26152


***

После проделанного анализа ясно видно, что вольшинство работодателей ищут людей в болших городах СНГ. Скорее всего поэтому средняя указанная ЗП чуть выше чем средняя по СНГ. Хотя в основном в вакансиях не указывают зарплатную вилку. Можем также наблюдать, что хоть и удаленная работа пользуется определенным спросом, все-таки абсолютное большинство вакансий предлагают полный рабочий день с полной занятостью. На данный момент на рынке труда будет тяжело найти работу если искатель в начале своей карьеры. 

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [24]:
query_5_1 = f'''select distinct e.name, count(v.id)
                from vacancies v
                    join employers e on v.employer_id=e.id
                group by e.name
                order by 2 desc
                limit 5
            '''

In [25]:
cnt_vac= pd.read_sql_query(query_5_1, connection)
cnt_vac.head()

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\4118035862.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  cnt_vac= pd.read_sql_query(query_5_1, connection)


,name,count
0,Яндекс,1933
1,Ростелеком,491
2,Тинькофф,444
3,СБЕР,428
4,Газпром нефть,331


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [26]:
query_5_1 = f'''select a.name as name, count(e.id) as emp, count(v.id) as vac
                from areas a
                    left join vacancies v on a.id=v.area_id
                    left join employers e on a.id= e.area
                group by a.name 
                having count(v.id)=0
                order by 2 desc
                limit 1
            '''

In [27]:
count_emp_vac= pd.read_sql_query(query_5_1, connection)
display(count_emp_vac)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\3544299062.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_emp_vac= pd.read_sql_query(query_5_1, connection)


,name,emp,vac
0,Россия,410,0


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [28]:
query_5_2 = f'''select e.name as name, count(distinct v.area_id) as vac
                from employers e
                    left join vacancies v on v.employer_id= e.id
                group by e.name
                order by 2 desc
            '''

In [29]:
emp_area= pd.read_sql_query(query_5_2, connection)
display(emp_area)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\310555112.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  emp_area= pd.read_sql_query(query_5_2, connection)


,name,vac
0,Яндекс,181
1,Ростелеком,152
2,Спецремонт,116
3,Поляков Денис Иванович,88
4,ООО ЕФИН,71
...,...,...
23170,СДЕЛКА,0
23171,Alandr Group,0
23172,СДК,0
23173,Lemon Land Lombard,0


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [30]:
query_5_3 = f'''select count(e.id)
                from employers e
                    left join employers_industries ei on ei.employer_id= e.id
                where ei.industry_id is null
            '''

In [31]:
ind_n= pd.read_sql_query(query_5_3, connection)
display(ind_n)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\1073280717.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ind_n= pd.read_sql_query(query_5_3, connection)


,count
0,8419


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [32]:
query_5_4 = f'''select e.name
                from employers e
                    join employers_industries ei on e.id=ei.employer_id
                group by e.id, e.name
                having count(distinct ei.industry_id)=4
                order by 1
                limit 1 offset 2
            '''

In [33]:
emp_ind= pd.read_sql_query(query_5_4, connection)
display(emp_ind)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\1775747399.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  emp_ind= pd.read_sql_query(query_5_4, connection)


,name
0,2ГИС


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [34]:
query_5_5 = f'''select count(distinct ei.employer_id)
                from employers_industries ei
                    join industries i on i.id=ei.industry_id
                where i.name = 'Разработка программного обеспечения'
            '''

In [35]:
count_emp= pd.read_sql_query(query_5_5, connection)
display(count_emp)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\1324660682.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_emp= pd.read_sql_query(query_5_5, connection)


,count
0,3553


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [36]:
import requests # Импортируем библиотеку requests
from bs4 import BeautifulSoup # Импортируем библиотеку BeautifulSoup

In [37]:
# Данное решение нашел в телеграм канале, поскольку перепробовал кучу решений и у меня ничего не вышло.
headers= {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/143.0.0.0 Safari/537.36'}
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8' # Определяем адрес страницы
response = requests.get(url, headers=headers)
page = BeautifulSoup(response.text, 'html.parser')
tabl = page.find("table", class_="standard")

millionniki = []
table = tabl.find("tbody")
rows = table.find_all("tr")


for i in rows:
    tags= list(i.find("a"))
    millionniki.append(tags[0])
millionniki= millionniki[1:]
print(millionniki)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']


In [38]:
query_5_6 = f'''select a.name as name, count(v.id) as cnt
                from vacancies v
                    join employers e on e.id=v.employer_id
                    join areas a on a.id = v.area_id
                where e.name = 'Яндекс' and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')
                group by a.name
                
                union all
                select 'Total' as name, count(v.id) as cnt
                from vacancies v
                    join employers e on e.id=v.employer_id
                    join areas a on a.id = v.area_id
                where e.name = 'Яндекс' and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')
                group by e.name
                order by 2 asc
            '''

In [39]:
yandex_mil= pd.read_sql_query(query_5_6, connection)
display(yandex_mil)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\3295075723.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  yandex_mil= pd.read_sql_query(query_5_6, connection)


,name,cnt
0,Омск,21
1,Челябинск,22
2,Красноярск,23
3,Волгоград,24
4,Пермь,25
5,Казань,25
6,Ростов-на-Дону,25
7,Уфа,26
8,Самара,26
9,Краснодар,30


***

Проведя анализ работодателей можно смело сказать, что компани в которых IT сектор развит нанимают больше всего сотрудников. Примером тому является компания Яндекс которая ищет 1933 специалиста в 181 регионе СНГ. Соит заметить, что несмотря на то что компании ищут людей живущих в конкретных регионах. Об этом мы узнали благодаря тому что в регионе Россия больше всего компаний но нет вакансий, это может говорить о непопулярности удаленной работы.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [40]:
query_6_1 = f'''select count(v.id)
                from vacancies v
                where lower(v.name) like '%data%' or lower(v.name) like '%данн%'                    
            '''

In [41]:
data_cnt= pd.read_sql_query(query_6_1, connection)
display(data_cnt)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\4191512999.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data_cnt= pd.read_sql_query(query_6_1, connection)


,count
0,1771


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [42]:
query_6_2 = f'''select count(v.id)
                from vacancies v 
                where (
                    lower(v.name) like '%data scientist%' or 
                    lower(v.name) like '%data science%' or 
                    lower(v.name) like '%исследователь данных%' or 
                    (v.name like '%ML%' and v.name not like '%HTML%') or 
                    lower(v.name) like '%machine learning%' or 
                    lower(v.name) like '%машинн%обучен%'
                ) and (
                    lower(v.name) like '%junior%' or 
                    v.experience= 'Нет опыта' or 
                    v.employment= 'Стажировка'
                )                   
            '''

In [43]:
data_sci_cnt= pd.read_sql_query(query_6_2, connection)
data_sci_cnt.head()

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\205585809.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data_sci_cnt= pd.read_sql_query(query_6_2, connection)


,count
0,51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [44]:
query_6_3 = f'''select count(v.id)
                from vacancies v 
                where (
                    lower(v.name) like '%data scientist%' or 
                    lower(v.name) like '%data science%' or 
                    lower(v.name) like '%исследователь данных%' or 
                    (lower(v.name) like '%ml%' and lower(v.name) not like '%html%') or 
                    lower(v.name) like '%machine learning%' or 
                    lower(v.name) like '%машинн%обучен%'
                ) and (
                    lower(v.key_skills) like '%postgres%' or
                    lower(v.key_skills) like '%sql%'
                )                   
            '''

In [45]:
sql_cnt= pd.read_sql_query(query_6_3, connection)
display(sql_cnt)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\2613755694.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  sql_cnt= pd.read_sql_query(query_6_3, connection)


,count
0,229


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [46]:
query_6_4 = f'''select count(v.id)
                from vacancies v 
                where (
                    lower(v.name) like '%data scientist%' or 
                    lower(v.name) like '%data science%' or 
                    lower(v.name) like '%исследователь данных%' or 
                    (lower(v.name) like '%ml%' and lower(v.name) not like '%html%') or 
                    lower(v.name) like '%machine learning%' or 
                    lower(v.name) like '%машинн%обучен%'
                ) and (
                    lower(v.key_skills) like '%python%'
                )                   
            '''

In [47]:
py_cnt= pd.read_sql_query(query_6_4, connection)
display(py_cnt)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\1545349579.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  py_cnt= pd.read_sql_query(query_6_4, connection)


,count
0,357


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [48]:
query_6_5 = f'''select *
                from vacancies v 
                where (
                    lower(v.name) like '%data scientist%' or 
                    lower(v.name) like '%data science%' or 
                    lower(v.name) like '%исследователь данных%' or 
                    (v.name like '%ML%' and v.name not like '%HTML%') or 
                    lower(v.name) like '%machine learning%' or 
                    lower(v.name) like '%машинн%обучен%'
                )                  
            '''

In [49]:
import numpy as np
avg_skills= pd.read_sql_query(query_6_5, connection)
def get_qty(text):
    x=0
    if text is not None:
        text= text.split('\t')
        x+=len(text)
        return x
    else:
        return np.nan
avg_skills['qty']=avg_skills['key_skills'].apply(get_qty)
round(avg_skills['qty'].mean(), 2)  

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\2437790653.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_skills= pd.read_sql_query(query_6_5, connection)


np.float64(6.41)

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [50]:
query_6_6 = f'''select v.experience as experience, avg((coalesce(v.salary_from, 0)+ coalesce(v.salary_to, 0))/
                    (case when v.salary_from is not null and v.salary_to is not null then 2 else 1 end)) as average_salary
                from vacancies v 
                where (
                    lower(v.name) like '%data scientist%' or 
                    lower(v.name) like '%data science%' or 
                    lower(v.name) like '%исследователь данных%' or 
                    (v.name like '%ML%' and v.name not like '%HTML%') or 
                    lower(v.name) like '%machine learning%' or 
                    lower(v.name) like '%машинн%обучен%'
                ) and (
                    v.salary_from is not null or
                    v.salary_to is not null
                    )
                group by v.experience                  
            '''

In [51]:
avg_salary= pd.read_sql_query(query_6_6, connection)
display(avg_salary)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\1391840803.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_salary= pd.read_sql_query(query_6_6, connection)


,experience,average_salary
0,Нет опыта,74642.857143
1,От 1 года до 3 лет,139674.750000
2,От 3 до 6 лет,243114.666667


***

Благодаря данному анализу мы можем наблюдать, что работа с данными крайне востребована, хотя для новичков порог вхождения в данную среду ощутимо велик. Особенно это касается data scientists для которых и вакансий сильно меньше и начальные средние  ЗП практически в 2 раза меньше. Одни из самых востребованныъ навыков- это SQL и Python. 

# Общий вывод по проекту

In [52]:
query_bonus = f'''select a.name as areas, i.name as industries, count(v.id) as vac
                from vacancies v
                    join areas a on v.area_id= a.id
                    join employers e on e.id= v.employer_id
                    join employers_industries ei on ei.employer_id= v.employer_id
                    join industries i on i.id= ei.industry_id
                group by a.name, i.name
                order by 3 desc                 
            '''

In [53]:
bonus= pd.read_sql_query(query_bonus,connection)
display(bonus)

C:\Users\alber\AppData\Local\Temp\ipykernel_29928\2907641292.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  bonus= pd.read_sql_query(query_bonus,connection)


,areas,industries,vac
0,Москва,Разработка программного обеспечения,1294
1,Москва,"Системная интеграция, автоматизации технологи...",1117
2,Минск,Разработка программного обеспечения,1023
3,Санкт-Петербург,Разработка программного обеспечения,908
4,Москва,"Интернет-компания (поисковики, платежные систе...",798
...,...,...,...
13532,Нахабино,Двигатели (производство),1
13533,Нахабино,"Железнодорожное машиностроение, вагоностроение",1
13534,Нахабино,Интернет-магазин,1
13535,Нахабино,"Курьерская, почтовая доставка",1


После проделанного анализа вакансий на hh.ru. мы можем увидеть что на рынке труда дефцит опытных IT специалистов. Большинство работадателей являются компании либо разрабатывающие програмное обеспечение, либо с сильно выделяющимся IT сектором(такие как банки например). Больше всего спроса на таких специалистов в столицах стран СНГ, где скорее всего находятся главные офисы компаний.